### PR TITLE
Fix #490, behaviour on num_sub_mixes

### LINKED ISSUE #490: Consider removing sub-mix loop
We currently only allow one sub-mix. We could remove the loops that define the mix presentation for multiple sub-mixes.

The existing mechanism allows for extensibility by adding multiple sub-mixes later, but the behavior would need to be strictly defined in order for old decoders to be compatible with newer bitstreams.

### PR DESCRIPTION
**Pull request by sunghee-hwang**
***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/pull/502.html" title="Last updated on Jun 28, 2023, 12:04 AM UTC (c7832ef)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/502/3318721...c7832ef.html" title="Last updated on Jun 28, 2023, 12:04 AM UTC (c7832ef)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1041,7 +1041,7 @@ class mix_presentation_obu() {
 
 <dfn noexport>mix_presentation_annotations()</dfn> is a class that provides informational metadata that an IA parser SHOULD refer to when selecting the [=Mix Presentation=] to use. The metadata MAY also be used by the playback system to display information to the user, but is not used in the rendering or mixing process to generate the final output audio signal.
 
-<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes.
+<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes. For this version of the specification, [=num_sub_mixes=] SHALL be set to 1. If a decoder or an OBU parser encounters a Mix Presentation OBU with [=num_sub_mixes=] > 1, it SHALL discard the Mix Presentation OBU.
```
Comment (sunghee-hwang) [inline index.bs:1044]:
```suggestion
<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes. For this version of the specification, [=num_sub_mixes=] SHALL be set to 1.  Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
```

**Review by cconcolato — APPROVED** [review]

File: index.bs [hunk 2]
```diff
@@ -1041,9 +1041,9 @@ class mix_presentation_obu() {
 
 <dfn noexport>mix_presentation_annotations()</dfn> is a class that provides informational metadata that an IA parser SHOULD refer to when selecting the [=Mix Presentation=] to use. The metadata MAY also be used by the playback system to display information to the user, but is not used in the rendering or mixing process to generate the final output audio signal.
 
-<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes. For this version of the specification, [=num_sub_mixes=] SHALL be set to 1.  Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
+<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes. It SHALL NOT be set to 0. For this version of the specification, it SHALL be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
```
Comment (cconcolato) [inline index.bs:1044]: You may want to code `num_sub_mixes_minus_1`?

File: index.bs [hunk 3]
```diff
@@ -1041,9 +1041,9 @@ class mix_presentation_obu() {
...
 
-<dfn noexport>num_audio_elements</dfn> specifies the number of [=Audio Element=]s that are used in this [=Mix Presentation=] to generate the final output audio signal for playback.
+<dfn noexport>num_audio_elements</dfn> specifies the number of [=Audio Element=]s that are used in this [=Mix Presentation=] to generate the final output audio signal for playback. It SHALL NOT be set to 0. For this version of the specification, it SHALL be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
```
Comment (cconcolato) [inline index.bs:1046]: It should be "> 2"

File: index.bs (hunk 2 above)
Comment (sunghee-hwang) [reply index.bs:1044]: > You may want to code `num_sub_mixes_minus_1`?

I would prefer a way to minimize the change of test bitstreams or semantics. There seems lots of syntaxes which we need to restrict such a thing.

File: index.bs (hunk 3 above)
Comment (sunghee-hwang) [reply index.bs]: addressed

File: index.bs [hunk 4]
```diff
@@ -1041,9 +1041,9 @@ class mix_presentation_obu() {
 
 <dfn noexport>mix_presentation_annotations()</dfn> is a class that provides informational metadata that an IA parser SHOULD refer to when selecting the [=Mix Presentation=] to use. The metadata MAY also be used by the playback system to display information to the user, but is not used in the rendering or mixing process to generate the final output audio signal.
 
-<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes.
+<dfn noexport>num_sub_mixes</dfn> specifies the number of sub-mixes. It SHALL NOT be set to 0. For this version of the specification, it SHALL be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
```
Comment (shawnisingh) [inline index.bs:1044]: Should this constraint (num_sub_mixes SHALL be 1) be associated with the profiles instead of being here?

File: index.bs [hunk 5]
```diff
@@ -1041,9 +1041,9 @@ class mix_presentation_obu() {
...
 
-<dfn noexport>num_audio_elements</dfn> specifies the number of [=Audio Element=]s that are used in this [=Mix Presentation=] to generate the final output audio signal for playback.
+<dfn noexport>num_audio_elements</dfn> specifies the number of [=Audio Element=]s that are used in this [=Mix Presentation=] to generate the final output audio signal for playback. It SHALL NOT be set to 0. For this version of the specification, it SHALL be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 SHALL be ignored by parsers compliant to this version of the specification.
```
Comment (shawnisingh) [inline index.bs:1046]: same here, this constraint makes more sense as a constraint associated with a specific profile?

File: index.bs [hunk 6]
```diff
@@ -1670,7 +1670,8 @@ Common restrictions on the [=IA Sequence=] for all profiles specified in this ve
 - In every [=Temporal Unit=], the start timestamp of every [=Audio Frame OBU=] SHALL be the same as that of every [=Parameter Block OBU=] if present.
 	- There SHALL be no redundant [=Parameter Block OBU=].
 	- [=Parameter Block OBU=]s shall come first and followed by [=Audio Frame OBU=]s. 	  
-- [=num_sub_mixes=] SHALL be set to 1.
+- [=num_sub_mixes=] SHALL be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHALL be ignored by parsers compliant to this version of the specification.
+- [=num_audio_elements=] SHALL be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 SHALL be ignored by parsers compliant to this version of the specification.
```
Comment (cconcolato) [inline index.bs:1674]: The sentences about the parsers ignoring certain values are in conflict with the sentences just before that mandate a certain value. Also the use of "this version of the specification" in the profile section is not correct.

If you expect that future streams compliant to this profile will contain num_audio_elements > 2 or num_sub_mixes > 1 then the SHALLs should be SHOULDs in the first sentences.

```suggestion
- [=num_sub_mixes=] SHOULD be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 MAY be ignored.
- [=num_audio_elements=] SHOULD be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 MAY be ignored.
NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to this new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.
```

File: index.bs [hunk 7]
```diff
@@ -1670,7 +1670,9 @@ Common restrictions on the [=IA Sequence=] for all profiles specified in this ve
 - In every [=Temporal Unit=], the start timestamp of every [=Audio Frame OBU=] SHALL be the same as that of every [=Parameter Block OBU=] if present.
 	- There SHALL be no redundant [=Parameter Block OBU=].
 	- [=Parameter Block OBU=]s shall come first and followed by [=Audio Frame OBU=]s. 	  
-- [=num_sub_mixes=] SHALL be set to 1.
+- [=num_sub_mixes=] SHOULD be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHOULD be ignored.
+- [=num_audio_elements=] SHOULD be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 SHOULD be ignored.
+NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to these new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.
```
Comment (sunghee-hwang) [inline index.bs:1675]:
```suggestion

NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to these new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.
```

Comment (tdaede) [inline index.bs:1675]:
```suggestion
  NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to these new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.
```

File: index.bs [hunk 8]
```diff
@@ -1670,7 +1670,9 @@ Common restrictions on the [=IA Sequence=] for all profiles specified in this ve
...
-- [=num_sub_mixes=] SHALL be set to 1.
+- [=num_sub_mixes=] SHOULD be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHOULD be ignored.
+- [=num_audio_elements=] SHOULD be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 SHOULD be ignored.
+  NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to these new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.
 - When [=num_layers=] = 1, DemixingParameDefinion() for demixng MAY be present in the [=Audio Element OBU=] and IA decoders MAY use the (default) parameter data for demixing for (dynamic) down-mixing.
 - All parameter definitons except DemixingParamDefinition() and ReconGainParamDefinition() SHALL have the same [=param_definition_mode=].
 	- When [=param_definition_mode=] = 0, [=duration=], [=num_subblocks=], [=constant_subblock_duration=] and [=subblock_duration=] SHALL be same in all parameter definitions, respectively.
```
Comment (sunghee-hwang) [inline index.bs:1678]:
```suggestion
- [=num_sub_mixes=] SHOULD be set to 1. Mix Presentation OBUs with [=num_sub_mixes=] > 1 SHOULD be ignored.
- [=num_audio_elements=] SHOULD be set to 1 or 2. Mix Presentation OBUs with [=num_audio_elements=] > 2 SHOULD be ignored.

NOTE: This behavior is to allow future versions of this specification to define new profiles that support a number of audio elements and/or a number of sub-mixes greater than those recommended in this profile, while still permitting streams compliant to these new profiles to be processed by parsers compliant to the profiles defined in this version of the specification.

- When [=num_layers=] = 1, DemixingParameDefinion() for demixng MAY be present in the [=Audio Element OBU=] and IA decoders MAY use the (default) parameter data for demixing for (dynamic) down-mixing.
- All parameter definitons except DemixingParamDefinition() and ReconGainParamDefinition() SHALL have the same [=param_definition_mode=].
	- When [=param_definition_mode=] = 0, [=duration=], [=num_subblocks=], [=constant_subblock_duration=] and [=subblock_duration=] SHALL be same in all parameter definitions, respectively.
```

**Pull request merged**